zilliqa/src/sync.rs

```diff
@@ -465,30 +585,190 @@ impl Sync {
     /// This will rebuild history from the previous marker, with another peer.
     /// If this function is called many times, it will eventually restart from Phase 0.
     fn retry_phase1(&mut self) -> Result<()> {
+        let SyncState::Retry1((range, marker)) = &self.state else {
+            unimplemented!("sync::RetryPhase1 : invalid state");
+        };
+
         self.retry_count = self.retry_count.saturating_add(1);
-        let segment_count = self.segments.count_sync_segments();
-        if segment_count == 0 {
-            tracing::error!("sync::Retry1 : no metadata segments");
-            self.state = SyncState::Phase0;
+        tracing::debug!(?range, "sync::Retry1 : retrying");
+
+        // Insert faux metadata - we only need the number, parent_hash
+        let mut faux_header = BlockHeader::genesis(Hash::ZERO);
+        faux_header.number = marker.number.saturating_add(1);
+        faux_header.qc.block_hash = marker.hash;
+
+        self.state = SyncState::Phase1(faux_header);
+        self.inject_at = None;
+
+        if Self::DO_SPECULATIVE {
+            self.do_sync()?;
+        }
+        Ok(())
+    }
+
+    /// Handle Passive Sync Request
+    pub fn handle_passive_request(
+        &mut self,
+        from: PeerId,
+        request: RequestBlocksByHash,
+    ) -> Result<ExternalMessage> {
+        tracing::debug!(hash = %request.hash, count = %request.count, %from,
+            "sync::PassiveRequest : received",
+        );
+
+        // Check if we should service this request - https://github.com/Zilliqa/zq2/issues/1878
+        if self.is_validator {
```

Comment:
Maybe its cleaner to just have a configuration variable which says whether you're willing to serve passing syncing responses?

Happy to change it later though.